diff -up libwmf-0.2.8.4/configure.ac.omv~ libwmf-0.2.8.4/configure.ac
--- libwmf-0.2.8.4/configure.ac.omv~	2018-05-14 21:37:19.777144936 +0200
+++ libwmf-0.2.8.4/configure.ac	2018-05-14 21:40:30.436658614 +0200
@@ -399,21 +399,10 @@ AC_ARG_WITH(freetype,[  --with-freetype=
 	fi
 ])
 
-if [ test -n "$FREETYPE_DIR" ]; then
-	AC_PATH_PROG(FREETYPE_CONFIG,freetype-config, ,[$FREETYPE_DIR/bin:$PATH])
-else
-	AC_PATH_PROG(FREETYPE_CONFIG,freetype-config)
-fi
-
-if [ test -n "$FREETYPE_CONFIG" ]; then
-	if [ test -n "$FREETYPE_DIR" ]; then
-		freetype_cflags="`$FREETYPE_CONFIG --cflags` -I$FREETYPE_DIR/include"
-		freetype_libs=`$FREETYPE_CONFIG --libs`
-	else
-		freetype_cflags=`$FREETYPE_CONFIG --cflags`
-		freetype_libs=`$FREETYPE_CONFIG --libs`
-	fi
-else
+PKG_CHECK_MODULES(FREETYPE,freetype2,[
+	freetype_cflags="`$PKG_CONFIG --cflags freetype2`"
+	freetype_libs="`$PKG_CONFIG --libs freetype2`"
+],[
 	if [ test -n "$FREETYPE_DIR" ]; then
 		freetype_cflags="-I$FREETYPE_DIR/include/freetype2 -I$FREETYPE_DIR/include"
 		freetype_libs="-L$FREETYPE_DIR/lib -lfreetype"
@@ -421,7 +410,7 @@ else
 		freetype_cflags=""
 		freetype_libs="-lfreetype"
 	fi
-fi
+])
 
 CPPFLAGS="$freetype_cflags $CPPFLAGS"
 LDFLAGS="$LDFLAGS $freetype_libs"
