--- libwmf-0.2.7/libwmf-config.in.fpons	2002-06-18 08:58:31.000000000 +0200
+++ libwmf-0.2.7/libwmf-config.in	2002-10-31 15:09:11.000000000 +0100
@@ -95,9 +95,6 @@
 	if test "$lib_gd" = "yes"; then
 		includes="$includes -I@includedir@/libwmf/gd"
 	fi
-	if test "$lib_wmf" = "yes"; then
-		includes="$includes -I@includedir@"
-	fi
 	echo $includes
 fi
 
@@ -114,5 +111,5 @@
 		fi
 	done
 
-	echo $libdirs $my_wmf_libs
+	echo $my_wmf_libs
 fi      
